storybookjs/storybook
# UI: Improve sidebar performance when switching stories

## What I did

3 changes that add up to a moderate speed-up when switching stories, especially in projects where there are many (several hundred) stories.
1. Fuse and list within the Search component are now only calculated when needed (when the value of the search input changes), not on each render.
2. The value returned by LayoutProvider is now memoized. This has a moderate impact because this hook is used in Node within the sidebar, which may have hundreds of instances. Before this change, useLayout was causing Node to rerender every time, making its React.memo useless.
3. The list of items inside Tree is now memoized. This is helpful because this list is made by mapping over a very large list
 
In sum, I saw a performance improvement of 10-40ms (12-50%) per render when changing stories. The impact will be more or less depending on how many stories you have.

<details><summary>Expand to see performance profiles 📈</summary>
Before:
<img width="1840" alt="storybook-perf-before-best" src="https://github.com/storybookjs/storybook/assets/20398475/8e2c56a8-9574-41de-a140-df25b4ed0801">
After:
<img width="1840" alt="storybook-perf-after-best" src="https://github.com/storybookjs/storybook/assets/20398475/5c1046a5-0fb0-4d0c-8af3-8444a17491eb">

[storybook-perf-profiles.zip](https://github.com/storybookjs/storybook/files/14394438/storybook-perf-profiles.zip) (unzip, then open JSON files with React Dev Tools)
</details> 

## Checklist for Contributors

### Testing



#### The changes in this PR are covered in the following automated tests:

(Unfortunately, I don't know of a good way to test this kind of performance improvements in React. )

#### Manual testing

1. To see the largest contrast in performance, choose any sandbox and create many stories. One way to do this quickly is by making a folder of tests with auto-generated names/titles and copy/paste the folder many times.
2. Open Storybook in your browser with React Dev Tools installed
3. Start a React profile, and switch between stories in the side bar


### Documentation

      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

(N/A)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release



This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_
